synapse/handlers/appservice.py

```diff
@@ -147,6 +169,60 @@ def query_room_alias_exists(self, room_alias):
                 defer.returnValue(result)
 
     @defer.inlineCallbacks
+    def query_3pu(self, protocol, fields):
+        services = yield self._get_services_for_3pn(protocol)
+
+        deferreds = []
+        for service in services:
+            deferreds.append(self.appservice_api.query_3pu(
+                service, protocol, fields
+            ))
+
+        results = yield defer.DeferredList(deferreds, consumeErrors=True)
+
+        ret = []
+        for (success, result) in results:
+            if not success:
+                continue
+            if not isinstance(result, list):
+                continue
+            for r in result:
+                if _is_valid_3pentity_result(r, field="userid"):
+                    ret.append(r)
+                else:
+                    logger.warn("Application service returned an " +
+                                "invalid result %r", r)
+
+        defer.returnValue(ret)
+
+    @defer.inlineCallbacks
+    def query_3pl(self, protocol, fields):
+        services = yield self._get_services_for_3pn(protocol)
+
+        deferreds = []
+        for service in services:
+            deferreds.append(self.appservice_api.query_3pl(
+                service, protocol, fields
+            ))
+
+        results = yield defer.DeferredList(deferreds, consumeErrors=True)
+
+        ret = []
+        for (success, result) in results:
+            if not success:
+                continue
+            if not isinstance(result, list):
+                continue
+            for r in result:
+                if _is_valid_3pentity_result(r, field="alias"):
+                    ret.append(r)
+                else:
+                    logger.warn("Application service returned an " +
+                                "invalid result %r", r)
+
+        defer.returnValue(ret)
```

Comment:
There seems to be a large from above. Can this be factored out somehow? Possibly making a function that returns 2-tuple of lists of aliases and user ids?